File: bob/Scripts/DeviceSelectDefault.js
/* This script sets the value of Device Selection combobox widget. Useful
   when the main OPI is opened with DEVICE='valid_default_device' macro.
   
   Without this script the OPI with a specified DEVICE macro would still
   work, but the combobox widget would be empty and the user perhaps
   slightly confused. This is purely for cosmetics, but it's nice to have. */

importPackage(Packages.org.csstudio.opibuilder.scriptUtil);

var device_macro = widget.getMacroValue("DEVICE");
var pv = widget.getPV();

if (widget.getValue() == "" && device_macro != null) {
	/* This sets the value of the widget, but not the PV associated with it. */
	widget.setValue(device_macro);
	
	/* It's important to set the value of loc://device PV to the same value. Otherwise the
	   local PV can trigger a widget value update with it's default value, which is "0.0". */
	pv.setValue(device_macro);
}

File: bob/Scripts/DeviceSelectDefault.py
# This script sets the value of Device Selection combobox widget. Useful
# when the main OPI is opened with DEVICE='valid_default_device' macro.
#   
# Without this script the OPI with a specified DEVICE macro would still
# work, but the combobox widget would be empty and the user perhaps
# slightly confused. This is purely for cosmetics, but it's nice to have. */

from org.csstudio.display.builder.runtime.script import ConsoleUtil
from org.csstudio.display.builder.runtime.script import ScriptUtil

device_macro = widget.getEffectiveMacros().getValue("DEVICE")
writable = False
pval = ""

try:
    writable = widget.getPropertyValue("pv_writable")
    pname = widget.getPropertyValue("pv_name")
    # Remove type spec and what follows
    pname = pname[:pname.find('<')]
    pv = ScriptUtil.getPVByName(widget, pname)
    pval = pv.getValue()
except Exception as e:
    msgout = "Exception in DeviceSelectDefault.py " + str(e)
    ConsoleUtil.writeInfo(msgout)

if (writable and (len(pval) <= 0) and (device_macro is not None)):
    widget.setPropertyValue("pv_value", device_macro)
    pv.setValue(device_macro)
